### ISSUE TITLE
Command-line: Wallet balance is not updated after sent transaction and fails on second tx

### ISSUE THREAD

**Issue by NiExTy** [spesmilo/electrum#1525]
I have wallet (restored from 1 private key) that is not updating balance after sent transactions.

**Querying wallet balance:**

``$ electrum getbalance``
```json
{
    "confirmed": "0.01"
}
```

**Sending transactions:**

``$ electrum paytomany ... | electrum broadcast -``
``
TXID
``

**Checking wallet balance again:**

``$ electrum getbalance``
```json
{
    "confirmed": "0.01"
}
```
:beetle: ``getbalance`` is returning old wallet balance, transaction is confirmed on blockchain.

``$ electrum getaddressbalance ADDRESS``
```json
{
    "confirmed": "0.0049", 
    "unconfirmed": "0"
}
```
``getaddressbalance`` returned correct address balance.

**Daemon status**

``$ electrum daemon status``
```json
{
    "auto_connect": true, 
    "blockchain_height": 381018, 
    "connected": true, 
    "nodes": [
        "electrum.bahiadelsol.io:5002:s", 
        "ecdsa.net:110:s", 
        "erbium1.sytes.net:50002:s", 
        "electrum.be:50002:s", 
        "VPS.hsmiths.com:50002:s", 
        "electrum.bauerj.eu:50002:s", 
        "us.electrum.be:50002:s"
    ], 
    "path": "/home/user2/.electrum", 
    "server": "ecdsa.net", 
    "server_height": 381018
}
```

Different servers and daemon restart didn't help on next transaction I am getting:

```python
Traceback (most recent call last):
  File "/usr/local/bin/electrum", line 473, in <module>
    result = run_command(config, None, password)
  File "/usr/local/bin/electrum", line 265, in run_command
    result = func(*args)
  File "/usr/local/lib/python2.7/dist-packages/electrum/commands.py", line 70, in func_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/electrum/commands.py", line 432, in paytomany
    tx = self._mktx(outputs, tx_fee, change_addr, domain, nocheck, unsigned)
  File "/usr/local/lib/python2.7/dist-packages/electrum/commands.py", line 414, in _mktx
    coins = self.wallet.get_spendable_coins(domain)
  File "/usr/local/lib/python2.7/dist-packages/electrum/wallet.py", line 617, in get_spendable_coins
    c = self.get_addr_utxo(addr)
  File "/usr/local/lib/python2.7/dist-packages/electrum/wallet.py", line 583, in get_addr_utxo
    coins.pop(txi)
KeyError: u'64-CHAR-HASH-EDITED-FOR-PRIVACY:4'
{
    "error": "Traceback (most recent call last):\n  File \"/usr/local/bin/electrum\", line 319, in run\n    response = run_command(config, self.network, password)\n  File \"/usr/local/bin/electrum\", line 265, in run_command\n    result = func(*args)\n  File \"/usr/local/lib/python2.7/dist-packages/electrum/commands.py\", line 70, in func_wrapper\n    return func(*args, **kwargs)\n  File \"/usr/local/lib/python2.7/dist-packages/electrum/commands.py\", line 228, in broadcast\n    return self.network.synchronous_get(('blockchain.transaction.broadcast', [str(t)]))\n  File \"/usr/local/lib/python2.7/dist-packages/electrum/transaction.py\", line 469, in __str__\n    self.raw = self.serialize()\n  File \"/usr/local/lib/python2.7/dist-packages/electrum/transaction.py\", line 651, in serialize\n    outputs = self.outputs\nAttributeError: Transaction instance has no attribute 'outputs'\n"
}
```
Version: Electrum 2.5.2

**Comment by ecdsa** [comment]
how did the traceback occur?

**Comment by NiExTy** [comment]
First transaction occurred without any error from command using ``paytomany``, second identical tx failed.

**Comment by ecdsa** [comment]
079cb311ecec0b42df6d28e22b9dd107cb5f6c2c should fix the first issue; not sure about the second one

**Comment by ecdsa** [comment]
ok, I just tested, it should fix both

**Comment by NiExTy** [comment]
Sent 2 transactions in a row, balance is updating fine, no problem here. After they confirmed, I tried to sent another tx and it is giving traceback error again.

So I am unable to send further transactions after initial wallet restore and after first transactions confirm. I have to recreate wallet after each transaction.

```python
Traceback (most recent call last):
  File "/usr/local/bin/electrum", line 5, in <module>
    pkg_resources.run_script('Electrum==2.5.2', 'electrum')
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 528, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1394, in run_script
    execfile(script_filename, namespace, namespace)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 474, in <module>
    result = run_command(config, None, password)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 266, in run_command
    result = func(*args)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 70, in func_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 432, in paytomany
    tx = self._mktx(outputs, tx_fee, change_addr, domain, nocheck, unsigned)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 414, in _mktx
    coins = self.wallet.get_spendable_coins(domain)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/wallet.py", line 612, in get_spendable_coins
    c = self.get_addr_utxo(addr)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/wallet.py", line 578, in get_addr_utxo
    coins.pop(txi)
KeyError: u'64-CHAR-HASH-EDITED-FOR-PRIVACY:19'
```

**Comment by NiExTy** [comment]
Now it hangs completely on ``paytomany``:

```python
Exception in thread Thread-12:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 319, in run
    self.client_pipe.send(response)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/util.py", line 438, in send
    out = json.dumps(request) + '\n'
  File "/usr/lib/python2.7/json/__init__.py", line 243, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <electrum.transaction.Transaction instance at 0xb44d4dac> is not JSON serializable
```

**Comment by ecdsa** [comment]
did you rstart the daemon after pulling the new code?

**Comment by NiExTy** [comment]
Stopped daemon, deleted ``.electrum`` folder and recreated wallet.

**Comment by ecdsa** [comment]
sorry abou that. it should work now

**Comment by NiExTy** [comment]
Sent 2 transactions and after they confirmed sent another one it is giving 2 tracebacks in a row:

```python
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 314, in run
    response = run_command(config, self.network, password)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 260, in run_command
    result = func(*args)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 70, in func_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 432, in paytomany
    tx = self._mktx(outputs, tx_fee, change_addr, domain, nocheck, unsigned)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 414, in _mktx
    coins = self.wallet.get_spendable_coins(domain)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/wallet.py", line 612, in get_spendable_coins
    c = self.get_addr_utxo(addr)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/wallet.py", line 578, in get_addr_utxo
    coins.pop(txi)
KeyError: u'64-CHAR-HASH-EDITED-FOR-PRIVACY:19:18'

Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 314, in run
    response = run_command(config, self.network, password)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/EGG-INFO/scripts/electrum", line 260, in run_command
    result = func(*args)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 70, in func_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/commands.py", line 228, in broadcast
    return self.network.synchronous_get(('blockchain.transaction.broadcast', [str(t)]))
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/transaction.py", line 469, in __str__
    self.raw = self.serialize()
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.2-py2.7.egg/electrum/transaction.py", line 651, in serialize
    outputs = self.outputs
AttributeError: Transaction instance has no attribute 'outputs'
```

**Comment by ecdsa** [comment]
that's weird..

**Comment by NiExTy** [comment]
Yeah, ``ecdsa.net`` is bugged for me and always fails after second confirmed transaction until server switch back-and-forth. Tried that multiple times with clean installs.

Maybe because ``ecdsa.net`` is _pruning_ (I have over >200 transactions on that address) or because it is running newer version than ``electrum.no-ip.org``?

**Comment by ecdsa** [comment]
does that issue also exist in the GUI ?

**Comment by NiExTy** [comment]
Yes, when it locked up in CLI, I started GUI and was getting the same error - send button was not working until I switched server back and forth, then it loaded history and send button was working again.

**Comment by ecdsa** [comment]
can you email me the bitcoin address that causes this?
also, can you try with another server that prunes with limit 100, like ecdsa.net?

**Comment by NiExTy** [comment]
I will send privkey shortly.
I tried another server with pruning limit 100 - ``VPS.hsmiths.com`` and it is also bugging. ``electrum.no-ip.org`` is still working fine.

**Comment by ecdsa** [comment]
thanks. having a look...

**Comment by ecdsa** [comment]
can you test the above fix? I pushed it in a separate branch, you have to:
`git pull`
`git checkout test1525`

**Comment by NiExTy** [comment]
Looks like working - didn't throw any traceback this time.

**Comment by ecdsa** [comment]
great. I will test a bit more before merging

**Comment by NiExTy** [comment]
Err, still coming back with 4th transaction.

**Comment by ecdsa** [comment]
should be fixed now (use master branch)

**Comment by NiExTy** [comment]
Still after second confirmed tx.

```python
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/EGG-INFO/scripts/electrum", line 314, in run
    response = run_command(config, self.network, password)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/EGG-INFO/scripts/electrum", line 260, in run_command
    result = func(*args)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/electrum/commands.py", line 70, in func_wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/electrum/commands.py", line 432, in paytomany
    tx = self._mktx(outputs, tx_fee, change_addr, domain, nocheck, unsigned)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/electrum/commands.py", line 414, in _mktx
    coins = self.wallet.get_spendable_coins(domain)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/electrum/wallet.py", line 612, in get_spendable_coins
    c = self.get_addr_utxo(addr)
  File "/usr/local/lib/python2.7/dist-packages/Electrum-2.5.3-py2.7.egg/electrum/wallet.py", line 578, in get_addr_utxo
    coins.pop(txi)
KeyError: u'64-CHAR-HASH-EDITED-FOR-PRIVACY:8'
```

**Comment by ecdsa** [comment]
ok, reopening.

**Comment by kyuupichan** [comment]
Can we start a new issue for this that describes the outstanding issue from the beginning and how to reproduce?  It's become very hard to follow.  
Thanks

**Comment by ecdsa** [comment]
better keep the discussion here, because the commits I made are relevant. 
I do not have a complete understanding of the issue.
it occurs with pruning. 
wallet.up_to_date seems to be set too early.

**Comment by kyuupichan** [comment]
OK.  Then can we have a full statement of the issue and how to reproduce?

**Comment by kyuupichan** [comment]
@ecdsa, my guess is this is something to do with a previously unpruned transaction (e.g. 100th on ecdsa.net) becoming the 101st or 102nd after the sends and therefore becoming pruned on the server and being removed from the server history sent to the client, and this causing an inconsistency issue on the client side.  Looking at the wallet code, the interdependent consistency requirements for txis, txos and histories are quite subtle; it would be good to remove the necessity of storing the redundant information (though I understand you did it for performance reasons).

Unfortunately because of the nature of this issue (an address with 100+ txs) and needing the ability to spend from that address this is hard to reproduce, but hopefully the above is not far off and a clue to what is going on.  I'll continue to think about what effects a tx dropping from history might have.

**Comment by kyuupichan** [comment]
Scratching my head further, I'm convinced this is an issue in wallet.py and its maintenance of its data structures, not the synchronizer.  Not entirely sure what, but this is suspicious:

receive_history_callback(), at the beginning when looping through items removed or changing height (as happens when confirmed), only calls remove_transaction conditionally (suppose the tx references more than one address in the wallet).  This might mean TXI and TXO become stale because add_transaction() will then not be called further down the function for that tx_hash.  I also note self.transactions is not updated for removed transactions.  

I cannot immediately see just looking at code (I have no testcase) how either of those could lead to a problem, although they look somewhat suspicious.

**Comment by NiExTy** [comment]
@kyuupichan Steps to reproduce:

```bash
# Clean up before start.
electrum daemon stop && sleep 2 && killall -9 electrum && rm -rf ~/.electrum
# Use server with low pruning limit (100) e.g. ecdsa.net or VPS.hsmiths.com.
electrum setconfig server VPS.hsmiths.com:50002:s
# Import privkey with transaction history over 100 (>250 in my case).
electrum restore :
electrum daemon start
electrum payto -f0.00002 ADDRESS 0.0001 | electrum brodcast -

# Wait while transaction confirms on blockchain and wallet updates (important).
electrum getbalance

# Repeat last send command after previous transaction is confirmed.
electrum payto -f0.00002 ADDRESS 0.0001 | electrum brodcast -
*EXCEPTION*
```
This happens only with servers with low pruning limit and addresses with long histories (just 1 input, but long previous history, probably more than pruning limit).

Also switching to full server and back was fixing error for next transaction, temporary only  - after that it fails again.

**Comment by kyuupichan** [comment]
Thanks, I try and figure this out

**Comment by kyuupichan** [comment]
When you're triggering the bug, are you paying back to the same address?  Or diff address with change sent back to orig address?

**Comment by NiExTy** [comment]
To different address with change sent back to origin address (default behaviour for imported key).

**Comment by kyuupichan** [comment]
Ok I've reproduced it.  I copied the wallet at the various stages through the process.  I'll try and debug it now and figure out what's happening.

**Comment by kyuupichan** [comment]
Out of curiosity - do you know if this happens if you do it all from the GUI?  i.e. does this only happen with the daemon?

**Comment by NiExTy** [comment]
Daemon only.

**Comment by kyuupichan** [comment]
Thanks.  I'm going to have to do this again.  I'll return the coins once finished testing.

**Comment by treodu** [comment]
I am having the same problem as the original post - only one difference - the wallet has never sent coins.  Isent coins to wallet which was created offline.  Blockchain.info shows the transaction and balance, but after seed import to electrum on an internet connected computer - electrum shows 0 (zero) balance.  Electrum CAN find transaction by transaction id, but balance is not updating.

**Comment by kyuupichan** [comment]
The problem happens when issuing the getbalance after the server serves history with the tx as confirmed.  Other steps seem fine.

I've also determined that if you do that getbalance with the wallet in the same state but a fresh daemon, it does not happen.

So it requires the daemon to have been running throughout.  Unfortunately I probably need to do one or two more runs, with more verbose logging to figure out what's really going on here.

**Comment by kyuupichan** [comment]
This is fascinating and very peculiar.  When the TX confirms, even though there is no client running, the daemon still gets notified (not unexpected) and there is still a Synchronizer running and printing debug messages about a subscription response for the address, and the wallet is saved.

This is unexpected (presumably some refs are circular and the syncronizer can't be eliminated?).  However I don't see it as a problem yet.  The TX that rolled off has its TXI, TXO and pruned_txo entries removed, as expected.  However, debug output shows the wallet as saved, but the wallet file isn't updated with these latter effects from the TX rolling off.

This is the current state of what I've figured out; still thinking / looking but posting this here in case @ecdsa has any ideas.

**Comment by kyuupichan** [comment]
OK I think I have it, we need to call save_transactions()

**Comment by kyuupichan** [comment]
I've committed branch 1525-fix.  @NiExTy - can you test it?  @ecdsa - it's not a great fix; what do you think?  See the commit comments that describe the issue and why I think this fixes it.

I have to go out now; sorry for dashing.  I'll be back online later.

**Comment by NiExTy** [comment]
Successfully sent multiple transactions using 1525-fix, not getting any traceback.

**Comment by ecdsa** [comment]
@kyuupichan thanks

**Comment by kyuupichan** [comment]
@NiExTy thanks, returned spent coins + fees

**Comment by ecdsa** [comment]
closing this for now. a proper fix will take more time

**Comment by treodu** [comment]
forgive me being a newbie but what is the magic solution so i can get correct balance in electrum?  thank you for all bitcoin users not on the inside of electrum development but who are greatly concerned for the safety/reliability/useability of this long dependable time vetted wallet.

**Comment by kyuupichan** [comment]
@treodu I doubt very much your issue has anything to do with this one.  If you're having a problem please open a new issue if you haven't already
